pkg/health/condition/check_backup_ready.go

```diff
@@ -56,53 +57,92 @@ func (a *backupReadyCheck) Check(ctx context.Context, etcd druidv1alpha1.Etcd) R
+		if err != nil {
+			return result
+		}
+
+		// Treat backup as succeeded if delta snap lease renewal happens in the required time window
+		// and full snap lease is not older than the computed full snapshot duration.
+		if time.Since(deltaLeaseRenewTime.Time) < 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
+			if time.Since(fullLeaseCreateTime.Time) < fullSnapshotDuration {
+				result.reason = BackupSucceeded
+				result.message = "Delta snapshot backup succeeded"
+				result.status = druidv1alpha1.ConditionTrue
+				return result
+			} else {
+				result.status = druidv1alpha1.ConditionFalse
+				result.reason = BackupFailed
+				result.message = "Full snapshot backup failed. Full snapshot lease created long ago, but not renewed"
+				return result
+			}
+		}
+
+	} else if fullLeaseRenewTime != nil && deltaLeaseRenewTime == nil {
+		// Most probable during a startup scenario for new clusters
+		// Special case: return Unknown condition for some time to allow delta backups to start up
+		// Even though full snapshot may have succeeded by the required time, we must still wait
+		// for delta snapshotting to begin to consider the backups as healthy, to maintain the given RPO
+		if time.Since(fullLeaseRenewTime.Time) < 5*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
+			result.message = "Waiting for periodic delta snapshotting to begin"
+			return result
+		} else {
+			result.status = druidv1alpha1.ConditionFalse
+			result.reason = BackupFailed
+			result.message = "Delta snapshot backup failed. Delta snapshot lease created long ago, but not renewed"
 			return result
 		}
-	} else if deltaLeaseRenewTime == nil && fullLeaseRenewTime != nil {
-		//Most probable during a startup scenario for new clusters
-		//Special case. Return Unknown condition for some time to allow delta backups to start up
-		if time.Since(fullLeaseRenewTime.Time) > 5*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
-			result.message = "Periodic delta snapshots not started yet"
+
+	} else {
+		// Both snap leases are renewed, ie, maintained. Both are expected to be renewed periodically
+		fullSnapshotDuration, err := utils.ComputeScheduleDuration(*etcd.Spec.Backup.FullSnapshotSchedule)
+		if err != nil {
 			return result
 		}
-	} else if deltaLeaseRenewTime != nil && fullLeaseRenewTime != nil {
-		//Both snap leases are maintained. Both are expected to be renewed periodically
-		if time.Since(deltaLeaseRenewTime.Time) < 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration && time.Since(fullLeaseRenewTime.Time) < 24*time.Hour {
+
+		// Mark backup as succeeded only if at least one of the last two delta snapshots has been taken successfully
+		// and if the last full snapshot has been taken according to the given full snapshot schedule
+		if time.Since(deltaLeaseRenewTime.Time) < 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration && time.Since(fullLeaseRenewTime.Time) < fullSnapshotDuration {
```

Comment:
@shreyas-s-rao ,

**In my opinion**:

- The present conditions seem to complicate the readability of the code. I propose simplifying these as in the below code snippet.
- Additionally, it appears we can leverage  `etcd.GetFullSnapshotLeaseName()` and `etcd.GetDeltaSnapshotLeaseName()` in place of `getFullSnapLeaseName` and `getDeltaSnapLeaseName`respectively. This change would allow us to remove these redundant functions.

What are your thoughts on this?

<details> <summary> <b>Please find the actual diff here </b> </summary>

```diff
diff --git a/pkg/health/condition/check_backup_ready.go b/pkg/health/condition/check_backup_ready.go
index 44451f5e..b3700bfe 100644
--- a/pkg/health/condition/check_backup_ready.go
+++ b/pkg/health/condition/check_backup_ready.go
@@ -23,6 +23,7 @@ import (
 	"github.com/gardener/etcd-druid/pkg/utils"
 	coordinationv1 "k8s.io/api/coordination/v1"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
@@ -42,105 +43,70 @@ const (
 	NotChecked string = "ConditionNotChecked"
 )
 
-func (a *backupReadyCheck) Check(ctx context.Context, etcd druidv1alpha1.Etcd) Result {
-	// Default case
-	result := &result{
-		conType: druidv1alpha1.ConditionTypeBackupReady,
-		status:  druidv1alpha1.ConditionUnknown,
-		reason:  Unknown,
-		message: "Cannot determine etcd backup status",
-	}
+func (a *backupReadyCheck) fetchLease(ctx context.Context, name string, namespace string) (*coordinationv1.Lease, error) {
+	lease := &coordinationv1.Lease{}
+	err := a.cl.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, lease)
+	return lease, err
+}
 
+func (a *backupReadyCheck) Check(ctx context.Context, etcd druidv1alpha1.Etcd) Result {
 	// Special case of etcd not being configured to take snapshots
 	// Do not add the BackupReady condition if backup is not configured
 	if etcd.Spec.Backup.Store == nil || etcd.Spec.Backup.Store.Provider == nil || len(*etcd.Spec.Backup.Store.Provider) == 0 {
 		return nil
 	}
 
-	// Fetch snapshot leases
-	var (
-		err            error
-		fullSnapLease  = &coordinationv1.Lease{}
-		deltaSnapLease = &coordinationv1.Lease{}
-	)
-
-	err = a.cl.Get(ctx, types.NamespacedName{Name: getFullSnapLeaseName(&etcd), Namespace: etcd.ObjectMeta.Namespace}, fullSnapLease)
+	fullSnapLease, err := a.fetchLease(ctx, etcd.GetFullSnapshotLeaseName(), etcd.Namespace)
 	if err != nil {
-		result.message = fmt.Sprintf("Unable to fetch full snap lease. %s", err.Error())
-		return result
+		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, fmt.Sprintf("Unable to fetch full snap lease. %s", err.Error()))
 	}
 
-	err = a.cl.Get(ctx, types.NamespacedName{Name: getDeltaSnapLeaseName(&etcd), Namespace: etcd.ObjectMeta.Namespace}, deltaSnapLease)
+	deltaSnapLease, err := a.fetchLease(ctx, etcd.GetDeltaSnapshotLeaseName(), etcd.Namespace)
 	if err != nil {
-		result.message = fmt.Sprintf("Unable to fetch delta snap lease. %s", err.Error())
-		return result
+		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, fmt.Sprintf("Unable to fetch delta snap lease. %s", err.Error()))
 	}
 
 	deltaLeaseRenewTime := deltaSnapLease.Spec.RenewTime
 	fullLeaseRenewTime := fullSnapLease.Spec.RenewTime
 	fullLeaseCreateTime := &fullSnapLease.ObjectMeta.CreationTimestamp
 
+	// Handle snapshot leases not yet renewed
 	if fullLeaseRenewTime == nil && deltaLeaseRenewTime == nil {
-		// Both snapshot leases are not yet renewed
-		result.message = "Snapshotter has not started yet"
-		return result
-
-	} else if fullLeaseRenewTime == nil && deltaLeaseRenewTime != nil {
-		// Most probable during reconcile of existing clusters if fresh leases are created
-
-		fullSnapshotDuration, err := utils.ComputeScheduleDuration(*etcd.Spec.Backup.FullSnapshotSchedule)
-		if err != nil {
-			return result
-		}
+		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, "Snapshotter has not started yet")
+	}
 
-		// Treat backup as succeeded if delta snap lease renewal happens in the required time window
-		// and full snap lease is not older than the computed full snapshot duration.
-		if time.Since(deltaLeaseRenewTime.Time) < 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
-			if time.Since(fullLeaseCreateTime.Time) < fullSnapshotDuration {
-				result.reason = BackupSucceeded
-				result.message = "Delta snapshot backup succeeded"
-				result.status = druidv1alpha1.ConditionTrue
-				return result
-			} else {
-				result.status = druidv1alpha1.ConditionFalse
-				result.reason = BackupFailed
-				result.message = "Full snapshot backup failed. Full snapshot lease created long ago, but not renewed"
-				return result
-			}
-		}
+	// Handle fresh leases
+	if fullLeaseRenewTime == nil && deltaLeaseRenewTime != nil {
+		return handleFreshLeases(&etcd, deltaLeaseRenewTime, fullLeaseCreateTime)
+	}
 
-	} else if fullLeaseRenewTime != nil && deltaLeaseRenewTime == nil {
-		// Most probable during a startup scenario for new clusters
-		// Special case: return Unknown condition for some time to allow delta backups to start up
-		// Even though full snapshot may have succeeded by the required time, we must still wait
-		// for delta snapshotting to begin to consider the backups as healthy, to maintain the given RPO
-		if time.Since(fullLeaseRenewTime.Time) < 5*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
-			result.message = "Waiting for periodic delta snapshotting to begin"
-			return result
-		} else {
-			result.status = druidv1alpha1.ConditionFalse
-			result.reason = BackupFailed
-			result.message = "Delta snapshot backup failed. Delta snapshot lease created long ago, but not renewed"
-			return result
-		}
+	// Handle startup scenario
+	if fullLeaseRenewTime != nil && deltaLeaseRenewTime == nil {
+		return handleStartupScenario(&etcd, fullLeaseRenewTime)
+	}
+	// Handle periodically maintained leases
+	return handleMaintainedLeases(&etcd, deltaLeaseRenewTime, fullLeaseRenewTime)
+}
 
-	} else {
-		// Both snap leases are renewed, ie, maintained. Both are expected to be renewed periodically
-		fullSnapshotDuration, err := utils.ComputeScheduleDuration(*etcd.Spec.Backup.FullSnapshotSchedule)
-		if err != nil {
-			return result
-		}
+// Handles the case where snapshot leases are periodically maintained
+func handleMaintainedLeases(etcd *druidv1alpha1.Etcd, deltaLeaseRenewTime, fullLeaseRenewTime *metav1.MicroTime) *result {
+	// Both snap leases are renewed, ie, maintained. Both are expected to be renewed periodically
+	fullSnapshotDuration, err := utils.ComputeScheduleDuration(*etcd.Spec.Backup.FullSnapshotSchedule)
+	if err != nil {
+		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, "Cannot compute full snapshot schedule duration")
+	}
 
-		// Mark backup as succeeded only if at least one of the last two delta snapshots has been taken successfully
-		// and if the last full snapshot has been taken according to the given full snapshot schedule
-		if time.Since(deltaLeaseRenewTime.Time) < 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration && time.Since(fullLeaseRenewTime.Time) < fullSnapshotDuration {
-			result.reason = BackupSucceeded
-			result.message = "Snapshot backup succeeded"
-			result.status = druidv1alpha1.ConditionTrue
-			return result
-		}
+	// Mark backup as succeeded only if at least one of the last two delta snapshots has been taken successfully
+	// and if the last full snapshot has been taken according to the given full snapshot schedule
+	if time.Since(deltaLeaseRenewTime.Time) < 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration && time.Since(fullLeaseRenewTime.Time) < fullSnapshotDuration {
+		return createBackupResult(druidv1alpha1.ConditionTrue, BackupSucceeded, "Snapshot backup succeeded")
 	}
 
+	return handleStaleSnapshotLeases(etcd, deltaLeaseRenewTime)
+}
+
+// Handles the case when snapshot leases are not updated for a long time
+func handleStaleSnapshotLeases(etcd *druidv1alpha1.Etcd, deltaLeaseRenewTime *metav1.MicroTime) *result {
 	// Cases where snapshot leases are not updated for a long time
 	// If snapshot leases are present and leases aren't updated, it is safe to assume that backup is not healthy
 	if etcd.Status.Conditions != nil {
@@ -154,24 +120,54 @@ func (a *backupReadyCheck) Check(ctx context.Context, etcd druidv1alpha1.Etcd) R
 		// Transition to "False" state only if present state is "Unknown" or "False"
 		if deltaLeaseRenewTime != nil && (prevBackupReadyStatus.Status == druidv1alpha1.ConditionUnknown || prevBackupReadyStatus.Status == druidv1alpha1.ConditionFalse) {
 			if time.Since(deltaLeaseRenewTime.Time) > 3*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
-				result.status = druidv1alpha1.ConditionFalse
-				result.reason = BackupFailed
-				result.message = "Stale snapshot leases. Not renewed in a long time"
-				return result
+				return createBackupResult(druidv1alpha1.ConditionFalse, BackupFailed, "Stale snapshot leases. Not renewed in a long time")
 			}
 		}
 	}
+	// Transition to "Unknown" state if we cannot prove a "True" state
+	return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, "Cannot determine etcd backup status")
+}
 
-	// Transition to "Unknown" state is we cannot prove a "True" state
-	return result
+// Handles the case of a startup scenario for new clusters
+func handleStartupScenario(etcd *druidv1alpha1.Etcd, fullLeaseRenewTime *metav1.MicroTime) *result {
+	// Special case: return Unknown condition for some time to allow delta backups to start up
+	// Even though full snapshot may have succeeded by the required time, we must still wait
+	// for delta snapshotting to begin to consider the backups as healthy, to maintain the given RPO
+	if time.Since(fullLeaseRenewTime.Time) < 5*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
+		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, "Waiting for periodic delta snapshotting to begin")
+	}
+	return createBackupResult(druidv1alpha1.ConditionFalse, BackupFailed, "Delta snapshot backup failed. Delta snapshot lease created long ago, but not renewed")
 }
 
-func getDeltaSnapLeaseName(etcd *druidv1alpha1.Etcd) string {
-	return fmt.Sprintf("%s-delta-snap", etcd.ObjectMeta.Name)
+// Handles the case of fresh leases being created
+func handleFreshLeases(etcd *druidv1alpha1.Etcd, deltaLeaseRenewTime *metav1.MicroTime, fullLeaseCreateTime *metav1.Time) *result {
+	// Most probable during reconcile of existing clusters if fresh leases are created
+	fullSnapshotDuration, err := utils.ComputeScheduleDuration(*etcd.Spec.Backup.FullSnapshotSchedule)
+	if err != nil {
+		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, "Cannot determine etcd backup status")
+	}
+
+	// Treat backup as succeeded if delta snap lease renewal happens in the required time window
+	// and full snap lease is not older than the computed full snapshot duration.
+	if time.Since(deltaLeaseRenewTime.Time) < 2*etcd.Spec.Backup.DeltaSnapshotPeriod.Duration {
+		if time.Since(fullLeaseCreateTime.Time) < fullSnapshotDuration {
+			return createBackupResult(druidv1alpha1.ConditionTrue, BackupSucceeded, "Delta snapshot backup succeeded")
+		} else {
+			return createBackupResult(druidv1alpha1.ConditionFalse, BackupFailed, "Full snapshot backup failed. Full snapshot lease created long ago, but not renewed")
+		}
+	}
+
+	return handleStaleSnapshotLeases(etcd, deltaLeaseRenewTime)
 }
 
-func getFullSnapLeaseName(etcd *druidv1alpha1.Etcd) string {
-	return fmt.Sprintf("%s-full-snap", etcd.ObjectMeta.Name)
+// Creates a backup result
+func createBackupResult(status druidv1alpha1.ConditionStatus, reason string, message string) *result {
+	return &result{
+		conType: druidv1alpha1.ConditionTypeBackupReady,
+		status:  status,
+		reason:  reason,
+		message: message,
+	}
 }
 
 // BackupReadyCheck returns a check for the "BackupReady" condition.

```

</details>


```go 
func (a *backupReadyCheck) Check(ctx context.Context, etcd druidv1alpha1.Etcd) Result {
	// Special case of etcd not being configured to take snapshots
	// Do not add the BackupReady condition if backup is not configured
	if etcd.Spec.Backup.Store == nil || etcd.Spec.Backup.Store.Provider == nil || len(*etcd.Spec.Backup.Store.Provider) == 0 {
		return nil
	}

	fullSnapLease, err := a.fetchLease(ctx, etcd.GetFullSnapshotLeaseName(), etcd.Namespace)
	if err != nil {
		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, fmt.Sprintf("Unable to fetch full snap lease. %s", err.Error()))
	}

	deltaSnapLease, err := a.fetchLease(ctx, etcd.GetDeltaSnapshotLeaseName(), etcd.Namespace)
	if err != nil {
		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, fmt.Sprintf("Unable to fetch delta snap lease. %s", err.Error()))
	}

	deltaLeaseRenewTime := deltaSnapLease.Spec.RenewTime
	fullLeaseRenewTime := fullSnapLease.Spec.RenewTime
	fullLeaseCreateTime := &fullSnapLease.ObjectMeta.CreationTimestamp

	// Handle snapshot leases not yet renewed
	if fullLeaseRenewTime == nil && deltaLeaseRenewTime == nil {
		return createBackupResult(druidv1alpha1.ConditionUnknown, Unknown, "Snapshotter has not started yet")
	}

	// Handle fresh leases
	if fullLeaseRenewTime == nil && deltaLeaseRenewTime != nil {
		return handleFreshLeases(&etcd, deltaLeaseRenewTime, fullLeaseCreateTime)
	}

	// Handle startup scenario
	if fullLeaseRenewTime != nil && deltaLeaseRenewTime == nil {
		return handleStartupScenario(&etcd, fullLeaseRenewTime)
	}
	// Handle periodically maintained leases
	return handleMaintainedLeases(&etcd, deltaLeaseRenewTime, fullLeaseRenewTime)
}
```

Reply:
Thanks @seshachalam-yv . So you agree with the changes to the actual conditions, but you would like to improve code readability, correct? That's definitely a valid point, and I'll make the changes. Thanks for pointing it out! 👍

Reply:
Overall conditions looks to me